sdk/appcenter-identity/src/main/java/com/microsoft/appcenter/identity/storage/PreferenceTokenStorage.java

```diff
@@ -32,61 +69,161 @@
+        List<TokenStoreEntity> history = getTokenHistory();
+        if (history == null || history.size() == 0) {
+            return new AuthTokenInfo(null, null, null);
+        }
+        TokenStoreEntity storeEntity = history.get(0);
+        String token = storeEntity.getToken();
+        Date endTime = storeEntity.getExpiresOn();
+        Date nextChangeTime = history.size() > 1 ? history.get(1).getTime() : null;
+        if (nextChangeTime != null && endTime != null && nextChangeTime.before(endTime)) {
+            endTime = nextChangeTime;
+        } else if (endTime == null && nextChangeTime != null) {
+            endTime = nextChangeTime;
         }
+        return new AuthTokenInfo(token, storeEntity.getTime(), endTime);
     }
 
     @Override
-    public void removeToken() {
-        SharedPreferencesManager.remove(PREFERENCE_KEY_AUTH_TOKEN);
-        SharedPreferencesManager.remove(PREFERENCE_KEY_HOME_ACCOUNT_ID);
-        AuthTokenContext.getInstance().clearToken();
+    public synchronized void removeToken(String token) {
+        List<TokenStoreEntity> history = getTokenHistory();
+        if (history == null || history.size() == 0) {
+            AppCenterLog.warn(LOG_TAG, "Couldn't remove token from history: token history is empty.");
+            return;
+        }
+        if (history.size() == 1) {
+            AppCenterLog.debug(LOG_TAG, "Couldn't remove token from history: token history contains only current one.");
+            return;
+        }
+        TokenStoreEntity storeEntity = history.get(0);
+        if (!TextUtils.equals(storeEntity.getToken(), token)) {
+            AppCenterLog.debug(LOG_TAG, "Couldn't remove token from history: the token isn't oldest or is already removed.");
+            return;
+        }
+
+        /* Remove the token from history. */
+        history.remove(0);
+        setTokenHistory(history);
+        AppCenterLog.debug(LOG_TAG, "The token has been removed from token history.");
+    }
+
+    @VisibleForTesting
+    List<TokenStoreEntity> getTokenHistory() {
+        if (mHistory != null) {
+            return mHistory;
+        }
+        String encryptedJson = SharedPreferencesManager.getString(PREFERENCE_KEY_TOKEN_HISTORY, null);
+        String json = null;
+        if (encryptedJson != null && !encryptedJson.isEmpty()) {
+            CryptoUtils.DecryptedData decryptedData = CryptoUtils.getInstance(mContext).decrypt(encryptedJson, false);
+            json = decryptedData.getDecryptedData();
+        }
+        if (json == null || json.isEmpty()) {
+            return null;
+        }
+        TokenStoreEntity[] entities = null;
+        try {
+            entities = new Gson().fromJson(json, TokenStoreEntity[].class);
+        } catch (JsonParseException e) {
+            AppCenterLog.warn(LOG_TAG, "Failed to deserialize auth token history.", e);
+        }
+        mHistory = entities != null ? new ArrayList<>(Arrays.asList(entities)) : new ArrayList<TokenStoreEntity>();
```

Comment:
Gson knows list and generics too, no need to use an intermediate array. Need to use the TypeToken syntax though.

Reply:
1. `TypeToken` 10x slower
2. List instead of array not work in test env (spent some time on it have no idea why exactly)